docs/developer-docs/smart-contracts/best-practices/idempotency.mdx

```diff
@@ -0,0 +1,90 @@
+
+Moreover, the ICP/ICRC ledgers use the `created_at_time` parameter to limit the validity period of a call. Roughly, the call is only considered valid if its `created_at_time` is not in the future and at most 24 hours in the past.[^ledger-created-at-details] This avoids the problem where the deduplication window expiring would allow a retried call to succeed again.
+
+But even with this improvement used in the ledgers, the time window approach implicitly assumes that the client will be able to get a definite answer to their call within the time window. For example, after the 24 hours expire, the user cannot easily tell if their ledger transfer happened; their only option is to analyze the ledger blocks, which is somewhat tedious, and has to be done carefully to avoid asynchrony issues; see the section on [queryable call results](#queryable-call-results).
+
+An alternative is to do away with the time window, and store the deduplication data forever. This requires multiple canisters to prevent exhausting the canister memory, similar to how the ICP/ICRC ledgers store the transaction data in the archive canister. This effectively moves the tedious part of querying the deduplication data (e.g., ledger blocks) from the user to the canister.
+
+Summarizing, the advantages of this approach are:
+1. It can support high throughput
+1. It requires no synchronization on part of the user, and supports use cases like multiple devices
+
+The disadvantages are:
+1. It is more complicated to implement than sequence numbers.
+1. If a time window is used, it usually implicitly assumes that the user learns the call outcome within the time window.
+1. The memory usage can grow fairly high with high supported throughput and long deduplication windows. For example, supporting 100 transactions per second with a deduplication window of 24 hours can require hundreds of megabytes of heap space. This can be mitigated by using multiple canisters to store the deduplication data, at the expense of further implementation complexity and higher latency.
+
+## Other approaches to safe retries
+
+In absence of idempotent endpoints, or even as an addition to them, clients may be able to use other endpoints to make their retries safe.
+
+### Queryable call results
+
+If the canister, in addition to the update endpoint, also exposes a query that can inform the user of the result of the update, the client can also use this for safe retries as follows:
+
+1. Attempt to perform the update
+1. If the result of the update is unknown (e.g., not present in the ingress history any more), query the call result endpoint to determine whether the call was applied or not. Moreover, one needs to ensure that the previously sent call call cannot be applied in the future. If both of these are true, the call might be retried, or safely reported as failed.
```

Comment:
```suggestion
1. If the result of the update is unknown (e.g., not present in the ingress history anymore), query the call result endpoint to determine whether the update was applied or not. Moreover, one needs to ensure that the previously sent call cannot be applied in the future. If both of these are true, the call might be retried, or safely reported as failed.
```

That being said, the paragraph is not particularly clear. What exactly does it mean for a call to not be applied in the future? If both of what are true?